include/seastar/core/chunked_fifo.hh

```diff
@@ -369,6 +378,29 @@ void chunked_fifo<T, items_per_chunk>::clear() noexcept {
 #endif
 }
 
+template <typename T, size_t items_per_chunk>
+void chunked_fifo<T, items_per_chunk>::pop_front_n(size_t n) noexcept {
+    while (n) {
```

Comment:
This method is going to be used by io_sink like this:

```c++
    size_t drained = 0;
    for (auto& rq : _pending_io) {
        if (!consume(rq)) {
            break;
        }
        drained++;
    }
    _pending_io.pop_front_n(drained);
```

IOW -- it first walks as many elements as it can (up to the whole container), then pop_front_n them in batch.

Would more classical approach be much worse:

```c++
    while (!_pending_io.empty()) {
        if (!consume(_pending_io.front()) {
            break;
        }
        _pending_io.pop_front();
    }
```

?

AFAICS complexity is going to be the same in big-O sense, the only difference seem to be that chunk->begin is incremented step-by-step in the latter case and "in batch" in the former

Reply:
Complexity is the same yes. As you say, however, I expect there is a difference in constant factor. The existing code was written using the bulk erase at the end, so I followed that pattern as I expected "it might be slower" to be the primary concern here and I wanted to minimize that. 

Here are some relevant results from the benchmark:

```
test                                             iterations      median         mad         min         max      allocs       tasks        inst      cycles
container_perf.pop_all_chunked_fifo               338190000     1.406ns     0.030ns     1.366ns     1.475ns       0.000       0.000        12.3         6.0
container_perf.pop_all_cicular_buffer             516880000     1.290ns     0.015ns     1.272ns     1.315ns       0.000       0.000        10.0         5.6
container_perf.erase_trivial_chunked_fifo         299160000     0.182ns     0.005ns     0.178ns     0.201ns       0.000       0.000         0.5         0.2
container_perf.erase_trivial_circular_buffer      715410000     0.143ns     0.005ns     0.137ns     0.153ns       0.000       0.000         0.1         0.1
container_perf.iter_big_chunked_fifo              640540000     0.822ns     0.003ns     0.814ns     0.825ns       0.000       0.000         9.0         3.6
container_perf.iter_big_circular_buffer          1240200000     0.479ns     0.002ns     0.477ns     0.483ns       0.000       0.000         8.0         2.1
```

`pop_all` is the approach you suggested with an "empty" consume, like so:

```
template <typename Traits, typename T>
size_t pop_all_bench(size_t size) {
    auto c = make_n<Traits, T>(size);

    perf_tests::start_measuring_time();
    while (!c.empty()) {
        perf_tests::do_not_optimize(c.front());
        c.pop_front();
    }
    perf_tests::stop_measuring_time();

    perf_tests::do_not_optimize(c);

    return size;
}
```

erase_trivial erases elements using either `pop_front_n` or in the case of `cicular_buffer`, `erase(b, e)`. 

All times are "per element". So `pop_front()` approach is 1.4 ns, whereas we expect the existing one to be something like the iteration time of 0.8 + erase time of 0.2, so ~1.0, pretty close.

Maybe simpler is better? Then the question is if `pop_front_n` should just be removed.

Reply:
I think it's fine either way. In the end the kernel is going to incur a large number of cache misses / branch mispredicts / spaghetti code issuing the queued up requests.

We should focus on higher level optimizations, like polling more frequently when we have many items, switching to io_uring (currently a small regression), and using IORING_SETUP_IOPOLL/IORING_SETUP_SQPOLL (and the other minor micro-optimizations) to run permanently without interrupts.

Reply:
> We should focus on higher level optimizations, like polling more frequently when we have many items, switching to io_uring (currently a small regression), and using IORING_SETUP_IOPOLL/IORING_SETUP_SQPOLL (and the other minor micro-optimizations) to run permanently without interrupts.

Thanks for the guidance, it makes a lot of sense and I agree with this approach.